### ISSUE TITLE
changing user of promulgated charm can cause revision to go down

### ISSUE THREAD

**Issue by rogpeppe** [juju/charmstore-client#203]
_From @kwmonroe on April 19, 2018 21:8_

I was under the impression that promulgated `cs:<charm>` revisions would monotonically increase, regardless of the `cs:~user/<charm>` rev that it pointed to.  This appears not to be the case with `filebeat`.  I recently tried to change the promulgated version to point at ~filebeat-charmers instead of ~containers.

Before I did anything, the latest stable promulgated version was https://jujucharms.com/filebeat/14.  As you can see on that page, it's `By containers`.  Here's what I did next:

```bash
$ charm show filebeat id
id:
  Id: cs:filebeat-14
  Name: filebeat
  Revision: 14

## unpromulgate the ~containers version
$ bhttp put -j https://api.jujucharms.com/charmstore/v5/~containers/filebeat/promulgate Promulgated:=false

## promulgate the ~filebeat-charmers version
$ bhttp put -j https://api.jujucharms.com/charmstore/v5/~filebeat-charmers/filebeat/promulgate Promulgated:=true

## OH NO!  The revision has gone from 14 down to 4
$ charm show filebeat id
id:
  Id: cs:~filebeat-charmers/filebeat-4
  Name: filebeat
  Revision: 4  # <-- BOOOO
  User: filebeat-charmers
```

At this point, visiting the top level filebeat page show the new owner, but the revision has dropped down to 4: https://jujucharms.com/filebeat/

This could cause problems for people that look at the page and think "oh, i'll just lock the latest filebeat rev in my bundle with `cs:filebeat-4`".  If they do that, they won't get the ~filebeat-charmers version -- they'll get a super old ~containers version: https://jujucharms.com/filebeat/4

Also confusing, on the current top level page (https://jujucharms.com/filebeat/), you'll see a link for `Latest version (#5)`.  Clicking that doesn't take you to ~filebeat-charmers/filebeat/5; it takes you to the super old ~containers version 5 (https://jujucharms.com/filebeat/5).

Is this by design, or has something with the promulgated revision incrementer broken?

_Copied from original issue: CanonicalLtd/charmstore-client#226_

**Comment by rogpeppe** [comment]
_From @kwmonroe on April 19, 2018 21:15_

FYI, the old ~charmers team uses `bhttp` to un/promulgate stuff.  If that may be causing this issue, please let me know the current way to handle promulgation.

**Comment by rogpeppe** [comment]
TL;DR: publish the ~filebeat-charmers/filebeat-15 to the stable channel and your woes will be behind you.

Long story (mostly so we can remember and understand the issue later)
------

In the charm store database, there are two main data structures: Entity and BaseEntity.
There is one BaseEntity for each ($user, $series, $name) tuple, and one
Entity for every uploaded charm.

The base entity holds ChannelEntities, a mapping from each channel
to the entity that's most recently published in that channel. When an
entity is published to a channel, the respective entry in EntityChannels
is updated accordingly. The base entity also holds whether a charm has
been promulgated or not.

Each entity records its fully qualified URL, and, if it's been
promulgated, the promulgated URL that it's been given.

To look up a promulgated charm name, such as "cs:filebeat", we find the
promulgated base entity with that name. Then we use ChannelEntities to
find the appropriate revision of the charm for the requested channel.

When we promulgate a charm, we update the promulgated URL for the latest
entity in the newly promulgated charm to be a URL with a revision one
more than the previous largest promulgated revision.

The problem in this case exhibits itself because the latest revision in
the newly promulgated charm is not published to the stable channel.

The sequence goes something like this:

	resolve "cs:filebeat"
		find base entity with name="filebeat", promulgated=true
		-> BaseEntity{
			URL: "cs:~filebeat-charmers/filebeat"
			ChannelEntities: {
				"stable": "cs:~filebeat-charmers/filebeat-4"
				"edge": "cs:~filebeat-charmers/filebeat-5"
			}
		}
		stable channel entity is cs:~filebeat-charmers/filebeat-4
		find entity with url="cs:~filebeat-charmers/filebeat-4"
		-> Entity{
			URL: "cs:~filebeat-charmers/filebeat-4"
			PromulgatedURL: ""
		}
	-> "cs:~filebeat-charmers/filebeat-4"

We end up with a charm entity that doesn't have a promulgated URL,
so we print the non-promulgated URL and revision instead.

To fix this issue, we need to make sure that when a charm is promulgated,
all its associated entities have promulgated URLs, so that when `charm release`
is called one any of those entities, it will have a promulgated URL.

A possible algorithm:

    for each revision of the newly promulgated charm:
        if there's no promulgated URL:
            set promulgated URL revision to one more than previous largest

This may have the undesired side-effect of suddenly increasing the
latest promulgated revision by a significant number.
